File: hw/medina/u3/hw33StrategyCode/TollsPython/src/main/python/ec/edu/espe/tools/utils/JsonPersistence.py
import os
import json
from ec.edu.espe.tools.utils.Persistence import Persistence
from ec.edu.espe.tools.model.Tool import Tool

class JsonPersistence(Persistence):
    FILE_NAME = "tools.json"

    def create(self, tool):
        tools = self.read()
        if any(t.getId() == tool.getId() for t in tools):
            return False
        tools.append(tool)
        return self._save_all(tools)

    def read(self):
        if not os.path.exists(self.FILE_NAME):
            return []
        try:
            raw = open(self.FILE_NAME, "r", encoding="utf-8").read().strip()
            if not raw:
                return []
            arr = json.loads(raw)
            tools = []
            for o in (arr or []):
                tools.append(Tool(o["id"], o["name"], float(o["price"]), o.get("materials", []) or [], float(o["priceWithIva"])))
            return tools
        except:
            return []

    def update(self, id, tool):
        tools = self.read()
        for i, t in enumerate(tools):
            if t.getId() == id:
                tools[i] = tool
                return self._save_all(tools)
        return False

    def delete(self, id):
        tools = self.read()
        new_tools = [t for t in tools if t.getId() != id]
        if len(new_tools) == len(tools):
            return False
        return self._save_all(new_tools)

    def find(self, id):
        for t in self.read():
            if t.getId() == id:
                return t
        return None

    def _save_all(self, tools):
        try:
            arr = [{
                "id": t.getId(),
                "name": t.getName(),
                "price": t.getPrice(),
                "materials": t.getMaterials() or [],
                "priceWithIva": t.getPriceWithIva()
            } for t in tools]
            with open(self.FILE_NAME, "w", encoding="utf-8") as f:
                json.dump(arr, f, indent=2)
            return True
        except:
            return False

File: hw/medina/u3/hw33StrategyCode/ToolsManagementJS_Localhost/src/main/javascript/ec/edu/espe/tools/utils/JsonPersistence.js
const fs = require("fs");
const path = require("path");
const Persistence = require("./Persistence");
const Tool = require("../model/Tool");

class JsonPersistence extends Persistence {
  static FILE_NAME = "tools.json";

  async create(tool) {
    const tools = await this.read();
    if (tools.some(t => t.getId() === tool.getId())) return false;
    tools.push(tool);
    return this._saveAll(tools);
  }

  async read() {
    const filePath = path.resolve(JsonPersistence.FILE_NAME);
    if (!fs.existsSync(filePath)) return [];
    const raw = fs.readFileSync(filePath, "utf-8");
    if (!raw.trim()) return [];
    try {
      const arr = JSON.parse(raw);
      return (arr || []).map(o => new Tool(o.id, o.name, o.price, o.materials || [], o.priceWithIva));
    } catch {
      return [];
    }
  }

  async update(id, tool) {
    const tools = await this.read();
    const idx = tools.findIndex(t => t.getId() === id);
    if (idx === -1) return false;
    tools[idx] = tool;
    return this._saveAll(tools);
  }

  async delete(id) {
    const tools = await this.read();
    const filtered = tools.filter(t => t.getId() !== id);
    if (filtered.length === tools.length) return false;
    return this._saveAll(filtered);
  }

  async find(id) {
    const tools = await this.read();
    return tools.find(t => t.getId() === id) || null;
  }

  _saveAll(tools) {
    try {
      const filePath = path.resolve(JsonPersistence.FILE_NAME);
      const plain = tools.map(t => ({
        id: t.getId(),
        name: t.getName(),
        price: t.getPrice(),
        materials: t.getMaterials(),
        priceWithIva: t.getPriceWithIva()
      }));
      fs.writeFileSync(filePath, JSON.stringify(plain, null, 2), "utf-8");
      return true;
    } catch {
      return false;
    }
  }
}

module.exports = JsonPersistence;
